zlib-rs/src/crc32.rs

```diff
@@ -0,0 +1,133 @@
+const CRC32_LSB_POLY: usize = 0xedb8_8320usize;
+const START_CHECKSUM: u32 = 0xffff_ffffu32;
+const FINAL_XOR: u32 = 0xffff_ffffu32;
+
+const W: usize = std::mem::size_of::<usize>();
+const N: usize = 5;
+
+static CRC32_BYTE_TABLE: [[u32; 256]; 1] = build_crc32_table(1);
+static CRC32_WORD_TABLE: [[u32; 256]; W] = build_crc32_table(1);
+static CRC32_STRIDE_TABLE: [[u32; 256]; W] = build_crc32_table(N);
+
+const fn build_crc32_table<const T: usize, const W: usize>(stride: usize) -> [[u32; T]; W] {
+    let mut arr = [[0u32; T]; W];
+    let mut i = 0;
+    while i < W {
+        let mut j = 0;
+        while j < T {
+            let mut c = j;
+            let mut k = 0;
+            while k < 8 * (W * stride - i) {
+                if c & 1 != 0 {
+                    c = CRC32_LSB_POLY ^ (c >> 1);
+                } else {
+                    c >>= 1;
+                }
+                k += 1;
+            }
+            arr[i][j] = c as u32;
+            j += 1;
+        }
+        i += 1;
+    }
+    arr
+}
+
+fn crc32_naive_inner(data: &[u8], start: u32) -> u32 {
+    data.iter().fold(start, |crc, val| {
+        let crc32_lsb = crc.to_le_bytes()[0];
+        CRC32_BYTE_TABLE[0][usize::from(crc32_lsb ^ *val)] ^ (crc >> 8)
+    })
+}
+
+pub fn crc32_naive(data: &[u8]) -> u32 {
+    crc32_naive_inner(data, START_CHECKSUM) ^ FINAL_XOR
+}
+
+fn crc32_words_inner(words: &[usize], start: u32, per_word_crcs: &[u32]) -> u32 {
+    words.iter().enumerate().fold(start, |crc, (i, word)| {
+        let mut word_crc = 0u32;
+        let value = *word ^ (crc ^ per_word_crcs.get(i).unwrap_or(&0)) as usize;
+        for (i, b) in value.to_le_bytes().into_iter().enumerate() {
+            word_crc ^= CRC32_WORD_TABLE[i][usize::from(b)];
+        }
+
+        // Per Listing 4 in Kadatch and Jenkins, this step
+        // is needed in case size_of(usize) < size_of(u32)
+        if std::mem::size_of_val(&word_crc) <= W {
+            word_crc
+        } else {
+            (crc >> 8) ^ word_crc
+        }
+    })
+}
+
+pub fn crc32_words(data: &[u8]) -> u32 {
```

Comment:
is this strictly a performance optimization?

Reply:
It was also useful for testing during development, but it could go away now that the rest is done.

`crc32_words_inner` is still needed because `crc32_interleaved` calls it.